source/common/ssl/context_impl.cc

```diff
@@ -503,9 +515,31 @@ bssl::UniquePtr<SSL> ClientContextImpl::newSsl() const {
     SSL_set_renegotiate_mode(ssl_con.get(), ssl_renegotiate_freely);
   }
 
+  if (max_session_keys_ > 0) {
+    absl::WriterMutexLock l(&session_keys_mu_);
+    if (!session_keys_.empty()) {
+      SSL_SESSION* session = session_keys_.front().get();
```

Comment:
Note, similarly to @julia-stripe's PR, this assumes a single session store per cluster, not per endpoint.

Reply:
@PiotrSikora can you talk a bit about the rationalization for that assumption?

Reply:
The assumption is wrong most of the time (i.e. we should save session per endpoint, not per cluster), which is one of the reasons why this was sitting in my local tree until now.

However, the per cluster store works perfectly fine for (a) single-endpoint clusters, (b) deployments using shared cache and/or session tickets, where the same session can be resumed across all endpoints, so it's a stepping stone in the right direction, and per endpoint store can be added a bit later.

Reply:
@PiotrSikora can you clarify what the next stones are towards per-endpoint stores?  I think @julia-stripe and I were under the impression there was a single `ClientContextImpl` per endpoint, rather than per-cluster, which it sounds like isn't true.  As you point out, this PR as stands won't work well for multi-endpoint clusters without shared session ticket keys, which is our big use-case.

Reply:
@bobby-stripe it's either:
1) Storing sessions in the endpoint object, so that session's lifetime is the same as endpoint's.
2) Adding ability to configure "session cache key", so that sessions stored in `ClientContextImpl` can be retrieved by a specific key(s) (e.g. `%UPSTREAM_CLUSTER%` - which is effectively what we have right now, `%UPSTREAM_HOST%` or `%REQ(:AUTHORITY)%`).

I'm leaning towards the second option, since it's much more flexible, but I didn't have time to work on this yet.

cc @julia-stripe @ggreenway

Reply:
@PiotrSikora how do you want to move forward with this? Merge this PR and then do 1/2 above? I'm guessing you get most of your wins by just doing (1) above.

Reply:
I'd leaning towards (2), since it's much more flexible solution.

But yeah, let's definitely merge this PR as-is (feature-wise).

Reply:
Can you open up a ticket for the per-endpoint continuation to track?